crypto/init.c

```diff
@@ -684,7 +706,47 @@ int OPENSSL_atexit(void (*handler)(void))
     return 1;
 }
 
-#ifdef OPENSSL_SYS_UNIX
+#ifndef OPENSSL_SYS_UNIX
+CRYPTO_RWLOCK *CRYPTO_THREAD_glock_new(const char *name)
+{
+    return CRYPTO_THREAD_lock_new();
+}
+#else
+
+
+/*
+ * Create a new global lock, return NULL on error.
+ */
+CRYPTO_RWLOCK *CRYPTO_THREAD_glock_new(const char *name)
+{
+    GLOBAL_LOCK *newlock;
+
+    if (name == NULL
+            || glock_lock == NULL
+            || (newlock = malloc(sizeof(*newlock))) == NULL)
+        return CRYPTO_THREAD_lock_new();
+    CRYPTO_THREAD_write_lock(glock_lock);
+    newlock->name = name;
+    newlock->lock = CRYPTO_THREAD_lock_new();
+    newlock->next = global_locks;
+    global_locks = newlock->next;
```

Comment:
??? I think you've meant `global_locks = newlock;`.

Reply:
BTW, it's why do you have to lock that early? Can CRYPTO_THREAD_lock_new() return error? One can also wonder if it's appropriate to expect that caller passes exclusively compile-time strings. I mean after all, `const char *` in argument list only means that callee won't change it, not that it has to be compile-time literal...

Reply:
I have to wonder why it's reinventing a linked list.

Reply:
the strings aren't ever touched.  They are just stored in case future debugging wants them (to track down deadlocks).  You're right the assignment was wrong.

I don't think I understand the "reinventing" comment.  This happens at init time, so I can't use any openssl facilities.

Reply:
> the strings aren't ever touched. They are just stored ...

No, pointers to them are stored, not strings. That I'm saying is what if caller passes dynamically allocated string and frees it. However, if you reply with "let's strdup" them, I'd rather suggest to reserve some storage as part of GLOBAL_LOCK, fixed length or as variable length extension [you know, malloc(sizeof(GLOBAL_LOCK) + strlen(name) + 1)]. Otherwise it should be separately and explicitly documented, than preferred name would be a compile-time literal.

Reply:
The function is not documented; I am not sure if we should?  The name field isn't used now either.